### PR TITLE
upgrade-2.x failure reporting

### PR DESCRIPTION
**Pull request by alexgg**
* Allow draft updates
* Use `raw_version` for image location fetch and fallback to `version` label
* Improve error reporting to cloud with the use of a signal trap handler

Relates-to: https://github.com/balena-os/meta-balena/issues/3257

### CODE REVIEW

**Comment by alexgg** [comment]
Manual test plan:

- On success:

- [ ] lock file is removed
- [ ] fifo files are removed
- [ ] supervisor is updated, target state version patched

- On interrupt, target state gets reported for:

- [ ] Interrupt before hostapp update
- [ ] Interrupt while hostapp update
- [ ] Interrupt before SV update
- [ ] Interrupt while SV update
- [ ] Interrupt after SV update

On error, target state gets reported for:

- [ ] Could not get the default supervisor version for this balenaOS release
- [ ] Could not get current supervisor version from the API
- [ ] Could not parse current supervisor version from the API
- [ ] Failed to patch supervisor version in target state
- [ ] all hostapp-update attempts have failed


- Supervisor update:

- [ ] On target state version patch error, HUP aborts and attempts to patch state to failed.


- Semver tests:

- [ ] New OS releases uses semver
- [ ] Older OS releases fallback to labels

- Regression test with balenahupr

**Comment by alexgg** [comment]
- Semver tests:

- [ ] New OS releases uses semver
- [ ] Older OS releases fallback to labels

- Regression test with balenahupr

**Comment by alexgg** [comment]
On error, target state gets reported for:

- [x] Could not get the default supervisor version for this balenaOS release
```
+ unset DEFAULT_SUPERVISOR_VERSION                                                                                                                                                                                                          
+ '[' -z '' ']'                                                                                                                                                                                                                             
+ log ERROR 'Could not get the default supervisor version for this balenaOS release, bailing out.'                                                                                                                                          
+ case $1 in                                                                                                                                                                                                                                
+ loglevel=ERROR                                                                                                                                                                                                                            
+ shift                                                                                                                                                                                                                                     
++ date +%s                                                                                                                                                                                                                                 
+ endtime=1694592947                                                                                                                                                                                                                        
+ '[' ERROR == ERROR ']'                                                                                                                                                                                                                    
+ printf '[%09d%s%s\n' 70 '][ERROR]' 'Could not get the default supervisor version for this balenaOS release, bailing out.'                                                                                                                 
[000000070][ERROR]Could not get the default supervisor version for this balenaOS release, bailing out.                                                                                                                                      
+ report_update_failed                                                                                                                                                                                                                      
+ perc=100                                                                                                                                                                                                                                  
+ state='OS update failed'                                                                                       
+ compare_device_state 100 'OS update failed'           
+ perc=100                                                                                                       
+ state='OS update failed'                              
+ local resp
```                 
- [x] Could not get current supervisor version from the API
```
+ unset CURRENT_SUPERVISOR_VERSION                         
+ '[' -z '' ']'                                                                                                                                                                                                                             + log ERROR 'Could not get current supervisor version from the API, got {"d":[{"supervisor_version":"14.10.10"}]}'
+ case $1 in                                               
+ loglevel=ERROR                                           
+ shift                                                    
++ date +%s                                                
+ endtime=1694601685                                       
+ '[' ERROR == ERROR ']'                                   
+ printf '[%09d%s%s\n' 70 '][ERROR]' 'Could not get current supervisor version from the API, got {"d":[{"supervisor_version":"14.10.10"}]}'
[000000070][ERROR]Could not get current supervisor version from the API, got {"d":[{"supervisor_version":"14.10.10"}]}
+ report_update_failed                                     
+ perc=100                                                 
+ state='OS update failed'                                 
+ compare_device_state 100 'OS update failed' 
```
- [x] Could not parse current supervisor version from the API
```
+ unset CURRENT_SUPERVISOR_VERSION                                                                                                                                                                                                          
+ '[' -n '' ']'                                                                                                                                                                                                                             
+ log ERROR 'Could not parse current supervisor version from the API (got {"d":[{"supervisor_version":"14.10.10"}]})...'                                                                                                                    
+ case $1 in                                                                                                                                                                                                                                
+ loglevel=ERROR                                                                                                                                                                                                                            
+ shift                                                                                                                                                                                                                                     
++ date +%s                                                                                                                                                                                                                                 
+ endtime=1694604200                                                                                                                                                                                                                        
+ '[' ERROR == ERROR ']'                                                                                                                                                                                                                    
+ printf '[%09d%s%s\n' 71 '][ERROR]' 'Could not parse current supervisor version from the API (got {"d":[{"supervisor_version":"14.10.10"}]})...'                                                                                           
[000000071][ERROR]Could not parse current supervisor version from the API (got {"d":[{"supervisor_version":"14.10.10"}]})...                                                                                                                
+ report_update_failed  
```
- [x] Failed to patch supervisor version in target state
```
[000000066][LOG]TEST: bailing out                                                                                                                                                                                                           
+ printf '[%09d%s%s\n' 66 '][LOG]' 'TEST: bailing out'                                                                                                                                                                                      
+ return 1                                                                                                                                                                                                                                  
+ log ERROR 'Failed to patch supervisor version in target state, bailing out.'                                                                                                                                                              
+ case $1 in                                                                                                                                                                                                                                
+ loglevel=ERROR                                                                                                                                                                                                                            
+ shift                                                                                                                                                                                                                                     
++ date +%s                                                                                                                                                                                                                                 
+ endtime=1694610870                                                                                                                                                                                                                        
+ '[' ERROR == ERROR ']'                                                                                                                                                                                                                    
+ printf '[%09d%s%s\n' 66 '][ERROR]' 'Failed to patch supervisor version in target state, bailing out.'                                                                                                                                     
[000000066][ERROR]Failed to patch supervisor version in target state, bailing out.                                                                                                                                                          
+ report_update_failed
```
- [x] all hostapp-update attempts have failed
```
+ printf '[%09d%s%s\n' 23 '][ERROR]' 'all hostapp-update attempts have failed...'                                                                                                                                                 [23/49320]
[000000023][ERROR]all hostapp-update attempts have failed...                                                                                                                                                                                
+ report_update_failed                                                                                                                                                                                                                      
+ perc=100                                                                                                                                                                                                                                  
+ state='OS update failed'                                                                                                                                                                                                                  
+ compare_device_state 100 'OS update failed'                                                                                                                                                                                               
+ perc=100                                                                                                                                                                                                                                  
+ state='OS update failed'                                                                                                                                                                                                                  
+ local resp                                                                                                                                                                                                                                
+ local remote_perc                                                                                                                                                                                                                         
+ local remote_state  
++ CURL_CA_BUNDLE=/tmp/tmp.2SVZli5jl3                                                                            
++ curl --silent --retry 10 --header 'Authorization: Bearer 1c7f057df9e1478a18ebefbaa747d6c0' 'https://api.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v6/device(uuid='\''a99efcce465bc94913e74bd3f8c9d960'\'')?$select=provisioning_state,pr
ovisioning_progress'                                    
++ jq '.d[]'                                                                                                     
+ resp='{                                                                                                        
  "provisioning_state": "Running OS update",                                                                     
  "provisioning_progress": 50                              
}'             
++ echo '{                                                                                                       
  "provisioning_state": "Running OS update",               
  "provisioning_progress": 50                              
}'     
++ jq -r .provisioning_progress                            
+ remote_perc=50    
++ echo '{              
  "provisioning_state": "Running OS update",                                                                     
  "provisioning_progress": 50                                                                                    
++ jq -r .provisioning_state                               
}'        
+ remote_state='Running OS update'                         
+ '[' -n 50 ']'                                         
+ '[' -n 'Running OS update' ']'                           
+ test 100 -eq 50         
+ (( c++ )) 
+ resin-device-progress --percentage 100 --state 'OS update failed'                                                   
+ continue                                                 
+ compare_device_state 100 'OS update failed'              
+ perc=100                                                 
+ state='OS update failed'                                 
+ local resp                                               
+ local remote_perc                                        
+ local remote_state                                       
++ jq '.d[]'                                               
++ CURL_CA_BUNDLE=/tmp/tmp.2SVZli5jl3                      
++ curl --silent --retry 10 --header 'Authorization: Bearer 1c7f057df9e1478a18ebefbaa747d6c0' 'https://api.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v6/device(uuid='\''a99efcce465bc94913e74bd3f8c9d960'\'')?$select=provisioning_state,pr
ovisioning_progress'                                       
+ resp='{                                                  
  "provisioning_state": "OS update failed",                
  "provisioning_progress": 100                             
}'                                          
```
- [ ] On failed SV update, device reboots into new OS and SV is updated on boot
```
[000000073][LOG]Successfully set supervision version in target state                                                                                                                                                                        
+ '[' LOG == ERROR ']'                                                                                                                                                                                                                      
+ printf '[%09d%s%s\n' 73 '][LOG]' 'Successfully set supervision version in target state'                                                                                                                                                   
+ rm -f /tmp/tmp.sAdgxmNX3Q                                                                                                                                                                                                                 
+ return 0                                                                                                                                                                                                                                  
+ _run_supervisor_update                                                                                                                                                                                                                    
+ local supervisor_update                                                                                                                                                                                                                   
+ local ret=0                                                                                                                                                                                                                               
+ return 1                                                                                                                                                                                                                                  
+ log WARN 'Failed to update supervisor to target state version - leave to first boot.'                                                                                                                                                     
+ case $1 in                                                                                                                                                                                                                                
+ loglevel=WARNING                                                                                                                                                                                                                          
+ shift                                                                                                                                                                                                                                     
++ date +%s                                                                                                                                                                                                                                 
+ endtime=1694602111                                                                                             
+ '[' WARNING == ERROR ']'
[000000073][WARNING]Failed to update supervisor to target state version - leave to first boot.                        
+ printf '[%09d%s%s\n' 73 '][WARNING]' 'Failed to update supervisor to target state version - leave to first boot.'
+ persistent_logging_config_var                            
++ CURL_CA_BUNDLE=/tmp/tmp.fQjimPAbwO                      
++ curl --retry 10 --silent -X GET 'https://api.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v5/device_config_variable?$filter=device/uuid%20eq%20'\''38fa3f0525b97104a532652e4f9a6b63'\''' -H 'Content-Type: application/json' -H 'Authorizat
ion: Bearer a07d3123dbbd7919409601193075c439'                                                                    
++ jq -r '.d[] | select((.name == "RESIN_SUPERVISOR_PERSISTENT_LOGGING") and (.value == "")) | .id'              
+ PROBLEMATIC_ENV_VAR=
+ '[' -n '' ']'                                         
+ finish_up registry2.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v2/f948ec85e12c717efa7768070c34a4a4@sha256:ca5a2474cfd127992c541ef1f9dd7a905dc091922ac83939e5b075afc4c2c4f2
+ update_package=registry2.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v2/f948ec85e12c717efa7768070c34a4a4@sha256:ca5a2474cfd127992c541ef1f9dd7a905dc091922ac83939e5b075afc4c2c4f2
+ '[' -n registry2.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v2/f948ec85e12c717efa7768070c34a4a4@sha256:ca5a2474cfd127992c541ef1f9dd7a905dc091922ac83939e5b075afc4c2c4f2 ']'
+ balena inspect registry2.06b63d5cda23ce87dd19c7aa56994e93.bob.local/v2/f948ec85e12c717efa7768070c34a4a4@sha256:ca5a2474cfd127992c541ef1f9dd7a905dc091922ac83939e5b075afc4c2c4f2
+ log 'No update package cleanup done'                                                                           
+ case $1 in          
+ loglevel=LOG 
++ date +%s                                                                                                      
+ endtime=1694602111                                                                                             
+ '[' LOG == ERROR ']'                                                                                           
+ printf '[%09d%s%s\n' 73 '][LOG]' 'No update package cleanup done'
+ sync                                                                                                           
[000000073][LOG]No update package cleanup done                                                                   
+ DOCKER_HOST=unix:///var/run/balena-host.sock                                                                   
+ balena logout registry2.06b63d5cda23ce87dd19c7aa56994e93.bob.local
+ '[' no == no ']'                                         
+ log 'Rebooting into new OS in 5 seconds...'              
+ case $1 in                                               
+ loglevel=LOG                                             
++ date +%s                                                
+ endtime=1694602111                                       
+ '[' LOG == ERROR ']'                                     
+ printf '[%09d%s%s\n' 73 '][LOG]' 'Rebooting into new OS in 5 seconds...'                                            
+ progress 100 'Update successful, rebooting'              
[000000073][LOG]Rebooting into new OS in 5 seconds...      
```
However, on newOS boot, the `update-balena-supervisor` script runs, the new image is downloaded, but the new
supervisor does not start. This needs to be investigated separately from this PR.

**Comment by alexgg** [comment]
- Semver tests:

- [ ] New OS releases uses semver
- [ ] Older OS releases fallback to labels

- Regression test with balenahupr

**Comment by alexgg** [comment]
- Semver tests:

For a RPI4 v2.98.33 is the first release that uses the `raw_version` API field instead of hostapp `version` labels.

- [ ] New OS releases uses semver
- [ ] Older OS releases fallback to labels

- Regression test with balenahupr

File: upgrade-2.x.sh [hunk 1]
```diff
@@ -992,9 +992,17 @@ LOGFILE="/mnt/data/balenahup/$SCRIPTNAME.$(date +"%Y%m%d_%H%M%S").log"
 mkdir -p "$(dirname "$LOGFILE")"
 echo "================$SCRIPTNAME HEADER START====================" > "$LOGFILE"
 date >> "$LOGFILE"
+
 # redirect all logs to the logfile, but also stderr to console (proxy)
-exec > >(cat >> "$LOGFILE")
-exec 2> >(tee -a "$LOGFILE" >&2)
+outfifo=$(mktemp -u)
+errfifo=$(mktemp -u)
+mkfifo "${outfifo}" "${errfifo}"
```
Comment (klutchell) [inline upgrade-2.x.sh:1033]: Very clever. I never thought to use a dry-run of mktemp to generate unique filenames for another command to create.

**Comment by klutchell** [comment]
Excellent changes @alexgg , no issues from my side!

**Pull request merged**